packages/eslint-plugin/src/rules/no-non-null-asserted-nullish-coalescing.ts

```diff
@@ -0,0 +1,94 @@
+import { TSESTree, TSESLint } from '@typescript-eslint/experimental-utils';
+import {
+  Definition,
+  DefinitionType,
+  Variable,
+} from '@typescript-eslint/scope-manager';
+import * as util from '../util';
+
+function hasAssignmentBeforeNode(
+  variable: Variable,
+  node: TSESTree.Node,
+): boolean {
+  return (
+    variable.references.some(
+      ref => ref.isWrite() && ref.identifier.range[1] < node.range[1],
+    ) ||
+    variable.defs.some(
+      def =>
+        isDefinitionWithAssignment(def) && def.node.range[1] < node.range[1],
+    )
+  );
+}
+
+function isDefinitionWithAssignment(definition: Definition): boolean {
+  if (definition.type !== DefinitionType.Variable) {
+    return false;
+  }
+
+  const variableDeclarator = definition.node;
+  return variableDeclarator.definite ?? variableDeclarator.init !== null;
+}
+
+export default util.createRule({
+  name: 'no-non-null-asserted-nullish-coalescing',
+  meta: {
+    type: 'problem',
+    docs: {
+      description:
+        'Disallows using a non-null assertion in the left operand of the nullish coalescing operator',
+      category: 'Possible Errors',
+      recommended: false,
+      suggestion: true,
+    },
+    messages: {
+      noNonNullAssertedNullishCoalescing:
+        'The nullish coalescing operator is designed to handle undefined and null - using a non-null assertion is not needed.',
+      suggestRemovingNonNull: 'You should remove the non-null assertion.',
+    },
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    return {
+      'LogicalExpression[operator = "??"] > TSNonNullExpression.left'(
+        node: TSESTree.TSNonNullExpression,
+      ): void {
+        if (node.expression.type === TSESTree.AST_NODE_TYPES.Identifier) {
+          const scope = context.getScope();
+          const identifier = node.expression;
+          const variable = scope.set.get(identifier.name);
```

Comment:
this is **almost** correct.
This will only get the variable if it's defined in the current scope:
```ts
function foo() {
  let x: string;
  x! ?? '';
}
```

But it won't handle this:
```ts
let x: string;
function foo() {
  x! ?? '';
}
```

To handle this you'll have to iteratively traverse `scope.upper` to check all scopes.

You should be able to use this utility to do this for you:

```ts
import {ASTUtils} from '@typescript-eslint/experimental-utils';

const variable = ASTUtils.findVariable(scope, identifier.name);
```